src/main/java/org/stellar/sdk/KeyPair.java

```diff
@@ -241,6 +250,34 @@ public DecoratedSignature signDecorated(byte[] data) {
     return decoratedSignature;
   }
 
+  /**
+   * Sign the provided payload data for payload signer where the input is the data being signed.
+   * Per the <a href="https://github.com/stellar/stellar-protocol/blob/master/core/cap-0040.md#signature-hint" CAP-40 Signature spec</a>
+   * {@link DecoratedSignature}.
+   *
+   * @param signerPayload the payload signers raw data to sign
+   * @return          DecoratedSignature
+   */
+  public DecoratedSignature signPayloadDecorated(byte[] signerPayload) {
+    DecoratedSignature payloadSignature = signDecorated(signerPayload);
+
+    byte[] hint = new byte[4];
+
+    // copy the last four bytes of the payload into the new hint
+    if (signerPayload.length >= hint.length) {
+      arraycopy(signerPayload, signerPayload.length - hint.length, hint, 0, hint.length);
+    } else {
+      arraycopy(signerPayload, 0, hint, 0, signerPayload.length);
+    }
+
+    //XOR the new hint with this keypair's public key hint
+    for (int i = 0; i < hint.length; i++) {
+      hint[i] ^= (i < payloadSignature.getHint().getSignatureHint().length ? payloadSignature.getHint().getSignatureHint()[i] : 0);
+    }
```

Comment:
❓ I'm a bit confused by this. It looks like there's a condition here to handle the `payloadSignature` having a too short hint, however the too short hint is handled a few lines above already. Hints are _always_ 4 bytes. A `SignatureHint` should never contain a hint shorter.

Reply:
`payloadSignature.getHint().getSignatureHint()` refers to the key hint, it was paranoid defensive guard on that key hint array being shorter than the payload hint array, fine with removing it for cleaner approach here.

Reply:
Yeah I think you can remove the check. Hint's should always be 4 bytes. In other languages where we can guarantee the length of the byte[] is 4 we do that, I guess in Java we can't?

Reply:
java array type doesn't include the dimension as part of the type declaration and the generated xdr.SignatureHint doesn't have any checks on the size. it would require decorator or other wrapper class in the sdk around xdr.SignatureHint that enforces the length dimension. I [removed that guard check](https://github.com/stellar/java-stellar-sdk/pull/417/commits/e6e196609085d48198931ac1e6d87c4e50bd5797).